### PR TITLE
fix: Fix time complexity when searching for common parent dir

### PR DESCRIPTION
**Pull request by SimonJang**
Small PR to fix the time complexity when searching for a common parent. 

This utility function is used when creating the dependency graph and when creating the fractal tree, so performance gains might be welcome there.

First PR here, so any comments on how to improve are more then welcome!

### CODE REVIEW

File: src/index/generateTrees/shared/findSharedParent.ts [hunk 1]
```diff
@@ -2,16 +2,19 @@ import path from "path";
...
-  firstParts.forEach((part, idx) => {
-    const allPartsMatch = parts.every(p => p[idx] === part);
+  const parentPaths: string[] = [];
+  const [shortestParts, secondShortestParts] = [
+    shortest,
+    secondShortest,
+  ].map(x => x.split("/"));
 
-    if (allPartsMatch) {
+  shortestParts.forEach((part, idx) => {
```
Comment (sQVe) [inline src/index/generateTrees/shared/findSharedParent.ts]: I'd prefer to return a filter here instead. That makes quite a bit more clear and readable.

File: src/index/generateTrees/shared/findSharedParent.ts [hunk 2]
```diff
@@ -4,17 +4,18 @@ import path from "path";
...
+  const parentPaths: string[] = [];
 
-    if (allPartsMatch) {
+  let index = 0;
+  for (const part of shortest.split("/")) {
+    if (part === secondShortestParts[index++]) {
       parentPaths.push(part);
     }
-  });
+  }
```
Comment (AnatoleLucet) [inline src/index/generateTrees/shared/findSharedParent.ts]: Why not using a `forEach`? Is it faster?

File: src/index/generateTrees/shared/findSharedParent.ts [hunk 3]
```diff
@@ -4,17 +4,18 @@ import path from "path";
...
-  const parentPaths: string[] = [];
-  const parts: string[][] = paths.map(x => x.split("/"));
-  const firstParts = parts[0];
+  const [shortest, secondShortest] =
+    paths.length > 2 ? paths.sort((a, b) => a.length - b.length) : paths;
 
-  firstParts.forEach((part, idx) => {
-    const allPartsMatch = parts.every(p => p[idx] === part);
+  const secondShortestParts = secondShortest.split("/");
+  const parentPaths: string[] = [];
```
Comment (sQVe) [inline src/index/generateTrees/shared/findSharedParent.ts]:
```suggestion
  return shortest
    .split("/")
    .filter((part, idx) => part === secondShortestParts[idx])
    .join("/");
```

File: src/index/generateTrees/shared/findSharedParent.ts [hunk 4]
```diff
@@ -4,17 +4,18 @@ import path from "path";
...
 
-    if (allPartsMatch) {
+  let index = 0;
+  for (const part of shortest.split("/")) {
```
Comment (SimonJang) [inline src/index/generateTrees/shared/findSharedParent.ts]: Change this this to `path.sep`

File: src/index/generateTrees/shared/findSharedParent.ts [hunk 5]
```diff
@@ -4,17 +4,18 @@ import path from "path";
...
 
-  const parentPaths: string[] = [];
-  const parts: string[][] = paths.map(x => x.split("/"));
-  const firstParts = parts[0];
+  const [shortest, secondShortest] =
+    paths.length > 2 ? paths.sort((a, b) => a.length - b.length) : paths;
 
-  firstParts.forEach((part, idx) => {
-    const allPartsMatch = parts.every(p => p[idx] === part);
+  const secondShortestParts = secondShortest.split("/");
```
Comment (SimonJang) [inline src/index/generateTrees/shared/findSharedParent.ts]: Change this this to `path.sep`

File: src/index/generateTrees/shared/findSharedParent.ts (hunk 2 above)
Comment (SimonJang) [reply src/index/generateTrees/shared/findSharedParent.ts]: Performance wise `for` loop is more interesting. https://jsperf.com/loops

Comment (sQVe) [reply src/index/generateTrees/shared/findSharedParent.ts]: @SimonJang I'd currently prefer readability before anything else currently. We do not have any noticeable performance issues.

Furthermore, there is a big difference in performance in a simple index based `for` loop and a `forof` loop. A `forof` loop is actually slower than a `forEach` loop.

**Comment by sQVe** [comment]
@SimonJang 🍻

**Pull request merged**